text/3476-expose-fn-type.md

```diff
@@ -0,0 +1,269 @@
+type MyTrait {
+    fn ambiguous();
+}
+impl MyTrait for MyStruct {
+    fn ambiguous() {  }
+}
+impl fn MyStruct::ambiguous { } // ERROR: ambiguous associated function
+// instead:
+impl fn <MyStruct as MyTrait>::ambiguous { } // OK
+```
+
+When the type of the associated function has generics, they will be handles as follows
+
+```rust
+struct MyStruct<T>(T);
+impl<T> MyStruct<T> {
+    fn get() -> T { .. }
+}
+
+impl<T> fn MyStruct::<T>::get { }
+// or fully described:
+impl<T> fn() -> T MyStruct::<T>::get { }
+```
+
+---
+
+When a function has generics, the function type is forced to be described, and the generic should be placed at it's desired position:
+```rust
+fn send<T: Send>(val: T, postal_code: u32) {}
+impl<T: Send> ParcelStation for fn(T, u32) send {
+    /* ... */
+}
+```
+
+When we have an implicit generic, the same rule applies
+```rust
+fn implicit_generic(val: impl Clone) -> impl ToString {}
+impl<T: Clone, U: ToString> for fn(T) -> U implicit_generic {
+    /* ... */
+}
+```
+
+---
+
+When functions have lifetimes, they have to be included in the types
+```rust
+fn log(text: &str) { .. }
+impl<'a> Logger for fn(&'a str) log {
+    /* ... */
+}
+```
+
+When the lifetime is explicitly defined on the function signature and there's no other rule forcing us to describe the function type, we can take a shortcut as follows
+```rust
+fn log<'a>(text: &'a str) { .. } // explicit lifetime 'a
+impl<'a> Logger for fn<'a> log {
+    /* ... */
+}
+```
+
+---
+
+Just as structs and enums have the possibility to derive traits to automatically generate code, function type have similar ways via attribute macros:
+
+```rust
+#[debug_signature]
+fn signature_test(val: i32) -> bool {
+    /* ... */
+}
+
+// Expands to
+
+fn signature_test(val: i32) -> bool {
+    /* ... */
+}
+impl DbgSignature for fn signature_test {
+    fn dbg_signature() -> &'static str {
+        "fn signature_test(val: i32) -> bool"
+    }
+}
+```
+
+Other than that, it should behave like every other type does.
+
+# Additional ToDo's
+
+## Change the fn type syntax for consistency
+
+When we try to compile the current [code snippet](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cba2a1391c3e431a7499c6bf427b350d) 
+```rust
+fn cool<'a, T: Clone>(val: &'a T) -> (i32, bool) {
+    todo!()
+}
+
+fn main() {
+    let _a: () = cool;
+}
+```
+we get following error:
+```
+error[E0308]: mismatched types
+ --> src/main.rs:6:18
+  |
+6 |     let _a: () = cool;
+  |             --   ^^^^ expected `()`, found fn item
+  |             |
+  |             expected due to this
+  |
+  = note: expected unit type `()`
+               found fn item `for<'a> fn(&'a _) -> (i32, bool) {cool::<_>}`
+
+For more information about this error, try `rustc --explain E0308`.
+```
+For consistency, we should change the syntax to `for<'a, T: Clone> fn(&'a T) -> (i32, bool) cool` (I'm not sure if we should put generics in the for)
+
+# Drawbacks
+[drawbacks]: #drawbacks
```

Comment:
It think it's worth mentioning that these named fn types could be used to name private return types and private arguments on public functions.

```rust
/// Workaround for `fn_traits` and/or `unboxed_closures` copied from
/// https://docs.rs/c/latest/src/never_say_never/lib.rs.html#1-77
mod fn_traits {
    pub trait FnOnce<Args> {
        type Output;
    }

    impl<F, R> FnOnce<()> for F
    where
        F: ::core::ops::FnOnce() -> R,
    {
        type Output = R;
    }
}

mod public {
    mod private {
        pub struct Private;
    }
    fn returns_private() -> private::Private {
        private::Private
    }
}

// We get to name `Private` by going through `returns_private`.
// This is only possible if we can name the functions type.
pub type Private = <fn public::returns_private as fn_traits::FnOnce<()>>::Output;
```

Accessing `Private` from the outer scope is not something that's possible today. This isn't something I'd mind having as private return types on public functions can be quite annoying :sweat_smile:. Fortunately type aliases can't be used as unit struct constructors. That would have made this type leak a much bigger issue.